xlive/H2MOD/Variants/GraveRobber/GraveRobber.cpp

```diff
@@ -0,0 +1,242 @@
+#include "stdafx.h"
+#include "GraveRobber.h"
+
+#include "H2MOD.h"
+#include "Blam/Engine/Game/GameGlobals.h"
+#include "H2MOD/Modules/CustomMenu/CustomLanguage.h"
+#include "H2MOD/Modules/HaloScript/HaloScript.h"
+#include "H2MOD/Modules/Shell/Config.h"
+#include "H2MOD/Utils/Utils.h"
+
+bool b_firstSpawn;
+const wchar_t* headhunterSoundTable[e_language_ids::_lang_id_end][e_graverobber_sounds::_graverobber_end]
+{
+	{SND_HEADHUNTER_EN, SND_SKULL_SCORED_EN},
+	{SND_HEADHUNTER_JP, SND_SKULL_SCORED_JP},
+	{SND_HEADHUNTER_GE, SND_SKULL_SCORED_GE},
+	{SND_HEADHUNTER_FR, SND_SKULL_SCORED_FR},
+	{SND_HEADHUNTER_ES, SND_SKULL_SCORED_ES},
+	{SND_HEADHUNTER_IT, SND_SKULL_SCORED_IT},
+	{SND_HEADHUNTER_KO, SND_SKULL_SCORED_KO},
+	{SND_HEADHUNTER_CH, SND_SKULL_SCORED_CH}
+};
+
+void GraveRobber::TriggerSound(e_graverobber_sounds sound, int sleep)
+{
+	const int language_id = *Memory::GetAddress<int*>(0x412818);
+
+	if (headhunterSoundTable[language_id][sound] != nullptr)
+	{
+		LOG_TRACE_GAME(L"[h2mod-graverobber] Triggering sound {}", headhunterSoundTable[language_id][sound]);
+		h2mod->custom_sound_play(headhunterSoundTable[language_id][sound], sleep);
+	}
+}
+
+void GraveRobber::SpawnPlayerClientSetup()
+{
+	if (b_firstSpawn)
+	{
+		TriggerSound(_snd_head_hunter, 1000);
+		b_firstSpawn = false;
+	}
+}
+
+void GraveRobber::SpawnSkull(datum unit_datum)
+{
+	const s_biped_data_definition* biped_unit = (s_biped_data_definition*)object_try_and_get_and_verify_type(unit_datum, FLAG(e_object_type::biped));
+
+	if (biped_unit != NULL)
+	{
+		s_object_placement_data nObject;
+
+		Engine::Objects::create_new_placement_data(&nObject, e_weapons_datum_index::ball, -1, 0);
+
+		nObject.position = biped_unit->position;
+		nObject.translational_velocity = biped_unit->translational_velocity;
+
+		datum new_object_datum = Engine::Objects::object_new(&nObject);
+		if (!DATUM_IS_NONE(new_object_datum))
+			Engine::Objects::simulation_action_object_create(new_object_datum);
+	}
+}
+
+typedef void(__thiscall* update_player_score_t)(void* thisptr, unsigned short a2, int a3, int a4, int a5, char a6);
+extern update_player_score_t p_update_player_score;
+
+void GraveRobber::PickupSkull(datum playerIdx, datum skullDatum)
+{
+	const short absPlayerIdx = DATUM_INDEX_TO_ABSOLUTE_INDEX(playerIdx);
+
+	if (!DATUM_IS_NONE(skullDatum))
+	{
+		typedef char* (__cdecl* get_score_data_ptr)();
+		auto p_get_score_data_ptr = Memory::GetAddress<get_score_data_ptr>(0x6B8A7, 0x6AD32);
+
+		char* player_score_data = p_get_score_data_ptr();
+		if (player_score_data)
+		{
+			if (!s_game_globals::game_is_predicted())
+			{
+				p_update_player_score(player_score_data, absPlayerIdx, 0, 1, -1, 0);
+			}
+			HaloScript::ObjectDestroy(skullDatum);
+
+			if (DATUM_INDEX_TO_ABSOLUTE_INDEX(h2mod->get_player_datum_index_from_controller_index(0)) == absPlayerIdx ||
+				DATUM_INDEX_TO_ABSOLUTE_INDEX(h2mod->get_player_datum_index_from_controller_index(1)) == absPlayerIdx ||
```

Comment:
we should create a helper function for this, also should be a loop

Reply:
https://github.com/pnill/cartographer/pull/407/commits/15ddad208effb18819e21dedf764dd9bbc93430a#diff-ac5bc43b1a79579fc36772ba42d52a8f5a4cd2c045d586ce2c21b4528b0e5d79
Changed it to a loop, and reworked the function so it doesn't have a bunch of nested if statements